### PR TITLE
Yaksh Learning Module

### LINKED ISSUE #385: Learning Module
Can run a course with lessons and quizzes.
Can have a hierarchical learning units.
Lesson unit can contain a lesson, exercise or a quiz(test).
There should be a way via which one can add videos to the lessons

### LINKED ISSUE #387: Learning Module Tests
Tests for the learning module

### PR DESCRIPTION
**Pull request by adityacp**
This PR adds a new feature of creating courses containing video lessons and quizzes.

Closes #385, Closes #387

### CODE REVIEW

File: requirements/requirements-common.txt [hunk 1]
```diff
@@ -6,4 +6,5 @@ pytz==2016.4
 python-social-auth==0.2.19
 selenium==2.53.6
 coverage
-ruamel.yaml==0.15.23
\ No newline at end of file
+ruamel.yaml==0.15.23
```
Comment (prabhuramachandran) [inline requirements/requirements-common.txt]: Why do we have hard requirements?  Why not `> 0.15` or so for all of these?

File: yaksh/models.py [hunk 2]
```diff
@@ -134,6 +140,346 @@ def create_trial_course(self, user):
     def get_hidden_courses(self, code):
         return self.filter(code=code, hidden=True)
 
+
+#############################################################################
+class Lesson(models.Model):
+    name = models.CharField(max_length=255)
+    description = models.TextField()
+    html_data = models.TextField(null=True, blank=True)
```
Comment (prabhuramachandran) [inline yaksh/models.py]: So where is the markdown saved?  Or is that in the description?  Perhaps a comment before the field may be useful for a reader?

File: yaksh/models.py [hunk 3]
```diff
@@ -134,6 +140,346 @@ def create_trial_course(self, user):
...
+
+#############################################################################
+class LessonFile(models.Model):
+    lesson = models.ForeignKey(Lesson, related_name="lesson")
+    file = models.FileField(upload_to=get_file_dir)
+
+    def remove(self):
+        if os.path.exists(self.file.path):
+            os.remove(self.file.path)
+            if os.listdir(os.path.dirname(self.file.path)) == []:
```
Comment (prabhuramachandran) [inline yaksh/models.py]: Just curious, why is this needed?

File: yaksh/models.py [hunk 4]
```diff
@@ -134,6 +140,346 @@ def create_trial_course(self, user):
...
+    description = models.CharField(max_length=256)
+
+    # Mininum passing percentage condition.
+    pass_criteria = models.FloatField("Passing percentage", default=40)
+
+    # Number of attempts for the quiz
+    attempts_allowed = models.IntegerField(default=1, choices=attempts)
+
+    time_between_attempts = models.IntegerField(
+        "Number of Days", choices=days_between_attempts
```
Comment (prabhuramachandran) [inline yaksh/models.py]: This is always hard to follow on the UI, change "Number of Days" to "Days between attempts", should this be the number of hours instead of days?

File: yaksh/models.py [hunk 5]
```diff
@@ -134,6 +140,346 @@ def create_trial_course(self, user):
...
+        desc = self.description or 'Quiz'
+        return '%s: on %s for %d minutes' % (desc, self.start_date_time,
+                                             self.duration)
+
+
+##########################################################################
+class LearningUnit(models.Model):
+    """ Maintain order of lesson and quiz added in the course """
+    order = models.IntegerField()
+    learning_type = models.CharField(max_length=16)
```
Comment (prabhuramachandran) [inline yaksh/models.py]: perhaps `type` is better than `learning_type`, also this should be a choice and not an arbitrary charfield.

File: yaksh/test_models.py [hunk 6]
```diff
@@ -1265,141 +1253,6 @@ def setUp(self):
             self.questions[3]
         )
 
-    def test_create_duplicate_course(self):
```
Comment (prabhuramachandran) [inline yaksh/test_models.py]: So why is this removed?

File: yaksh/models.py (hunk 2 above)
Comment (adityacp) [reply yaksh/models.py]: markdown is saved in the description and html_data contains converted HTML text. Will add comments.

File: yaksh/models.py (hunk 3 above)
Comment (adityacp) [reply yaksh/models.py]: This is to check if the directory is empty so as to delete the directory if the directory is empty.

File: yaksh/test_models.py (hunk 6 above)
Comment (adityacp) [reply yaksh/test_models.py]: Will be added again with new tests.

File: yaksh/models.py (hunk 5 above)
Comment (adityacp) [reply yaksh/models.py]: The user does not specifically create learning unit, it is our internal model. So learning type can be saved while adding quizzes and lesson to a module. So we don't need a choice field.

File: requirements/requirements-common.txt (hunk 1 above)
Comment (ankitjavalkar) [reply requirements/requirements-common.txt]: Specific pinning is usually recommended, so as to avoid bugs/errors arising out of incompatible dependencies.

**Comment by ankitjavalkar** [comment]
Please add a "Add New Course" buton to the ```yaksh/courses.html``` template. It is very unsettling for a user to have to go to the dashboard to create a new course but view the courses list on the Courses page.

File: yaksh/templates/yaksh/design_course_session.html [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        <div class="row">
+            <div class="col-md-8 col-md-offset-2 available-list">
+                <div id="fixed-available-wrapper">
+                    <p><u><b>Available Lessons and quizzes: (Add Lessons and Quizzes)</b></u></p>
+                    <div id="fixed-available">
+                        <table id="course-details" class="table table-bordered">
+                        <tr>
+                        <th width="2%">Select</th>
+                        <th>Learning Module</th>
+                        <th>Learning Units</th>
```
Comment (ankitjavalkar) [inline yaksh/templates/yaksh/design_course_session.html]: Can we please change this to "Lessons/Quizzes" instead of introducing new terminology

File: yaksh/models.py [hunk 8]
```diff
@@ -266,18 +593,25 @@ def remove_teachers(self, *teachers):
...
                                            creator=user)
             quiz = Quiz()
-            demo_quiz = quiz.create_demo_quiz(course)
+            demo_quiz = quiz.create_demo_quiz(user)
             demo_ques = Question()
             demo_ques.create_demo_questions(user)
             demo_que_ppr = QuestionPaper()
             demo_que_ppr.create_demo_quiz_ppr(demo_quiz, user)
             success = True
+            ordered_unit = LearningUnit.objects.create(
```
Comment (maheshgudi) [inline yaksh/models.py]: We could add a demo lesson as well

File: yaksh/models.py [hunk 9]
```diff
@@ -266,18 +593,25 @@ def remove_teachers(self, *teachers):
...
+                order=1, learning_type="quiz", quiz=demo_quiz)
+            learning_module = LearningModule.objects.create(
```
Comment (maheshgudi) [inline yaksh/models.py]: The description says none. It should also have html_data

File: yaksh/templates/yaksh/monitor.html [hunk 10]
```diff
@@ -46,7 +46,7 @@
     {% if course.get_quizzes %}
     <td>
     {% for quiz in course.get_quizzes %}
-     <li class="list-group-item"><a href = "{{URL_ROOT}}/exam/manage/monitor/{{quiz.id}}">
+     <li class="list-group-item"><a href = "{{URL_ROOT}}/exam/manage/monitor/{{quiz.id}}/{{course.id}}/">
```
Comment (maheshgudi) [inline yaksh/templates/yaksh/monitor.html]: Can we monitor students' progress

File: yaksh/views.py [hunk 11]
```diff
@@ -25,6 +25,7 @@
 import six
 from textwrap import dedent
 import zipfile
+from markdown import Markdown
```
Comment (maheshgudi) [inline yaksh/views.py]: markdown has an updated package called markdown2. You can try it.

File: yaksh/templates/yaksh/add_module.html [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                <div id="fixed-added-wrapper">
+                    <p><u><b>Choosen Lessons and quizzes:</b></u>
+                    </p>
+                    <div id="fixed-added">
+                        <table id="course-details" class="table table-bordered">
+                        <tr>
+                        <th>Select</th>
+                        <th>Quiz/Lesson</th>
+                        <th>Order</th>
+                        <th width="20%">Check Prerequisite
```
Comment (maheshgudi) [inline yaksh/templates/yaksh/add_module.html]: This should be more self-explanatory.

File: yaksh/templates/yaksh/show_video.html [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+<br><br><br>
+<center><h3>Lessons/Quizzes</h3></center>
+<ul class="list">
+{% for unit in learning_units %}
+<span>
+<li>
+{% get_unit_status course learning_module unit user as status %}
+
+{% if unit.id == current_unit.id %}
+    <span class="glyphicon glyphicon-edit" data-toggle="tooltip" title="Currently on">
```
Comment (maheshgudi) [inline yaksh/templates/yaksh/show_video.html]: This should be  glyphicon glyphicon-pencil. I suggest it would be more appropriate

File: .coveragerc [hunk 14]
```diff
@@ -6,10 +6,25 @@ omit =
...
+     manage.py
+     online_test/wsgi.py
+     online_test/__init__.py
+     yaksh/__init__.py
+     yaksh/pipeline/*
+     yaksh/migrations/*
+     yaksh/templatetags/__init__.py
+     yaksh/middleware/__init__.py
+     setup.py
+     tasks.py
```
[the remark this answers is not in the document]
Comment (adityacp) [reply .coveragerc]: Adding tests now will create a huge PR changes. We can add it later.

File: yaksh/forms.py [hunk 15]
```diff
@@ -311,8 +272,8 @@ def save(self, commit=True, *args, **kwargs):
 
     class Meta:
         model = Course
-        exclude = ['creator', 'requests', 'students', 'rejected',
-            'created_on', 'is_trial', 'hidden', 'teachers']
+        fields = ['name', 'enrollment', 'active', 'code', 'instructions',
```
[the remark this answers is not in the document]
Comment (adityacp) [reply yaksh/forms.py]: In either case, we might have to add field names. `exclude` creates a big list. So I felt it would be better to add field names.

File: yaksh/models.py (hunk 4 above)
Comment (adityacp) [reply yaksh/models.py]: Will be changed in a separate PR.

**Comment by adityacp** [comment]
Resolved most of the comments.
Created issue #411 for unresolved comments.
